File: packages/llm-obs-node-packages/shared-infra/src/http/pipeline/step-admission-control.ts
/**
 * @file step-admission-control.ts
 * @description Pipeline Step 1: Concurrency Admission Control & Load Shedding.
 */

import type { PipelineStep, PipelineContext } from "./types";
import { HTTP_CONSTANTS } from "../constants";

export class StepAdmissionControl implements PipelineStep {
  public readonly name = "AdmissionControl";
  public readonly description = "Concurrency Admission Control & Fleet Capacity Guard";

  public async execute<T>(ctx: PipelineContext<T>): Promise<void> {
    ctx.stepIndex++;
    if (!ctx.admissionControl.acquire()) {
      const activeCount = ctx.admissionControl.getActiveCount();
      ctx.span?.addEvent(HTTP_CONSTANTS.EVENT_EXECUTION_FAILURE, {
        "admission_control.shed": true,
        "active_count": activeCount,
      });
      throw new Error(`Too Many Requests - Fleet In-Flight Concurrency Capacity (${activeCount}) Exceeded`);
    }
  }
}

File: packages/llm-obs-python-packages/python-shared/src/python_shared/http/pipeline/step_admission_control.py
from python_shared.http.pipeline.types import PipelineContext

class StepAdmissionControl:
    name: str = "AdmissionControl"
    description: str = "Concurrency Admission Control & Fleet Capacity Guard"

    async def execute(self, ctx: PipelineContext) -> None:
        ctx.step_index += 1
        if not ctx.admission_control.acquire():
            active_count = ctx.admission_control.get_active_count()
            raise RuntimeError(f"Too Many Requests - Fleet In-Flight Concurrency Capacity ({active_count}) Exceeded")
